service-config/src/main/java/com/palantir/config/service/ServiceDiscoveryConfiguration.java

```diff
@@ -0,0 +1,121 @@
+import org.immutables.value.Value.Style;
+
+/**
+ * Configuration class that contains a map of {@code serviceName}s and their respective {@link ServiceConfiguration}s
+ * and provides utility methods for apps to build service clients.
+ */
+@Immutable
+@JsonDeserialize(as = ImmutableServiceDiscoveryConfiguration.class)
+@Style(visibility = Style.ImplementationVisibility.PACKAGE)
+public abstract class ServiceDiscoveryConfiguration {
+
+    @JsonProperty("apiToken")
+    public abstract Optional<BearerToken> defaultApiToken();
+
+    @JsonProperty("security")
+    public abstract Optional<SslConfiguration> defaultSecurity();
+
+    public abstract Map<String, ServiceConfiguration> services();
+
+    /**
+     * Uses the provided service name as the key to retrieve the API token for the respective
+     * {@link ServiceConfiguration}. If the API token is not defined in the {@link ServiceConfiguration}, the default
+     * API token will be returned.
+     *
+     * @param serviceName
+     *        the name of the service to be used to retrieve the respective {@link ServiceConfiguration}
+     * @returns the API token for a service based on the input service name
+     */
+    public final Optional<BearerToken> getApiToken(String serviceName) {
+        checkArgument(!Strings.isNullOrEmpty(serviceName));
+
+        Optional<ServiceConfiguration> maybeServiceConfig = getServiceConfiguration(serviceName);
+        if (!maybeServiceConfig.isPresent()) {
+            return defaultApiToken();
+        }
+
+        return maybeServiceConfig.get().apiToken().or(defaultApiToken());
+    }
+
+    /**
+     * Uses the service name as the key to retrieve the {@link SSLConfiguration} for the respective
+     * {@link ServiceConfiguration}. If the {@link SslConfiguration} is not defined in the {@link ServiceConfiguration},
+     * the default {@link SslConfiguration} will be returned.
+     *
+     * @param serviceName
+     *        the name of the service to be used to retrieve the respective {@link ServiceConfiguration}
+     * @returns the {@link ServiceConfiguration} for a service based on the input service name
+     */
+    public final Optional<SslConfiguration> getSslConfiguration(String serviceName) {
+        checkArgument(!Strings.isNullOrEmpty(serviceName));
+
+        Optional<ServiceConfiguration> maybeServiceConfig = getServiceConfiguration(serviceName);
+        if (!maybeServiceConfig.isPresent()) {
+            return defaultSecurity();
+        }
+
+        return maybeServiceConfig.get().security().or(defaultSecurity());
+    }
+
+    /**
+     * Uses the service name as the key to retrieve the respective {@link ServiceConfiguration}.
+     *
+     * @param serviceName
+     *        the name of the service.
+     * @returns the {@link ServiceConfiguration} for a service based on the input service name
+     */
+    public final Optional<ServiceConfiguration> getServiceConfiguration(String serviceName) {
```

Comment:
probably easier to return the nullable thing here, and let it throw RuntimeException when you try to use it later?